src/app.rs

```diff
@@ -224,7 +276,24 @@ impl App {
                         self.help_text(),
                     ));
                 }
-                None => self.help(),
+                None => match self.action_with_result {
+                    Some(action_with_result) => {
+                        if args.contains(&"-h".to_string()) || args.contains(&"--help".to_string())
+                        {
+                            self.help();
+                            return;
+                        }
+                        match action_with_result(&Context::new(
+                            args[1..].to_vec(),
+                            self.flags.clone(),
+                            self.help_text(),
+                        )) {
+                            Ok(_) => (),
+                            Err(e) => panic!("{}", e.message),
```

Comment:
Instead of calling `panic!` here, I think it would be better to return an error from this `run` function, which we might rename to `run_with_result`, and to make this a nonbreaking change, we could add a new `run` function that has the same interface as the current one and which calls `run_with_result` and panics if that returns an error. What do you think about this? If that sounds good to you, I could make that change in this PR or a followup (please let me know your preference). Thanks!

Reply:
@timburks 
I think that suggestion is good, please implement `run_with_result`.

Reply:
We would like to replace the `run` implementation with a `run_with_result` implementation in the future, as we want to prioritize simplicity.

Reply:
@ksk001100 Thanks for your feedback! Yes, I'll proceed with that in this PR.

Reply:
@ksk001100 I think `run_with_result` is ready to review. I've added some top-level tests that demonstrate calling it and I've put a comment in #75 showing how it can be used to return Result values directly from `main()`.

I agree with eventually fully replacing `run` with `run_with_result` (and just calling the new function `run`), and also agree with waiting for a major version bump to make that breaking change. I'm happy to make those changes whenever you'd like to do that.

Since this is a fairly large change, feel free to take your time to review and get feedback from anyone else before merging. Thank you!